subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpClientHelper.java

```diff
@@ -138,15 +142,48 @@ private HttpClientResponse performHttpRequest(HttpRequestBase request, HttpConte
     }
 
     private HttpClientResponse toHttpClientResponse(HttpRequestBase request, HttpContext httpContext, CloseableHttpResponse response) {
+        validateRedirectChain(httpContext);
         URI lastRedirectLocation = getLastRedirectLocation(httpContext);
         URI effectiveUri = lastRedirectLocation == null ? request.getURI() : lastRedirectLocation;
         return new HttpClientResponse(request.getMethod(), effectiveUri, response);
     }
 
-    @SuppressWarnings("unchecked")
-    private URI getLastRedirectLocation(HttpContext httpContext) {
-        List<URI> redirectLocations = (List<URI>) httpContext.getAttribute(REDIRECT_LOCATIONS);
-        return (redirectLocations == null || redirectLocations.isEmpty()) ? null : Iterables.getLast(redirectLocations);
+    private void validateUrl(URI url) {
+        /*
+         * For security purposes this intentionally requires a user to opt-in to http on case by case basis.
+         * We do not want to have a system/gradle property that allows a universal disable of this check.
+         */
+        if (settings.allowInsecureProtocol()) {
+            return;
+        }
+
+        final String scheme = url.getScheme();
+        if (!"https".equalsIgnoreCase(scheme)) {
+            DeprecationLogger.nagUserOfDeprecated("Uploading or resolving content over insecure protocol '" + scheme + "'");
```

Comment:
If there's a better nag method to use and a suggested advice I can give the user, please let me know.

Reply:
I would try something like this:
`nagUserWithDeprecatedIndirectUserCodeCause("Insecure HTTP requests", "Switch the protocol to HTTPS or allow insecure protocols.", "The URL was " + url)`

I don't believe it's a problem here, but we should be careful that `url` doesn't contain credentials.  e.g., someone adds `https://user:password@myserver` as a URL.

Reply:
I've added URI credentials stripping logic to all of our errors in `HttpClientHelper`.